libraries/reconstruction_cuda/cuda_gpu_iirconvolve.cu

```diff
@@ -1,107 +1,301 @@
-// inspired by https://devblogs.nvidia.com/efficient-matrix-transpose-cuda-cc/
-// TILE_DIM=32, BLOCK_ROWS=8
-// No bank-conflict transpose
-// Same as transposeCoalesced except the first tile dimension is padded
-// to avoid shared memory bank conflicts.
-// can be used to transpose non-square 2D arrays
-__global__
-void transposeNoBankConflicts32x8(float *odata, const float *idata, int xdim, int ydim) {
-    __shared__ float tile[32][32 + 1];
-    int tilex = blockIdx.x * 32;
-    int tiley = blockIdx.y * 32;
-    int x = tilex + threadIdx.x;
-    int y = tiley + threadIdx.y;
-
-    for (int j = 0; j < 32; j += 8) {
-        int index = (y + j) * xdim + x;
-        if (index < (xdim*ydim)) {
-            tile[threadIdx.y + j][threadIdx.x] = idata[index];
-        }
+#include "cuda_asserts.h"
+
+namespace iirConvolve2D_Cardinal_BSpline_3_MirrorOffBoundKernels {
+
+const int WARP_SIZE = 32;
+
+/*
+    Changing these parameters can affect speed on different devices
+    Values in curly brackets are sets of allowed values that can
+    be used
+    BLOCK_SIZE = Xdim and Ydim of block in row pass
+    COL_BLOCK_SIZe = Xdim of block in column pass
+*/
+
+const int BLOCK_SIZE = 64; // { 16, 32, 64 }
+const int COL_BLOCK_SIZE = 128; // { 32, 64, 128, 256, 512, 1024 } ... above 128 only if it is multiple of dim size
+const int UNROLL = 8;
+
+template< typename T >
+using shared_type = T[BLOCK_SIZE][BLOCK_SIZE+1];
+template< typename T >
+using data_ptr = T * __restrict__;
+template< typename T >
+using warp_shared_type = volatile T[WARP_SIZE];
+
+__device__ float power(float base, int exp) {
```

Comment:
See https://github.com/I2PC/xmipp/blob/devel/libraries/reconstruction_cuda/cuda_basic_math.h.
If it's not there, please add it